hashintel/hash
# H-1651: Persist user plugin settings in database

## 🌟 What is the purpose of this PR?

The user's settings in the browser plugin are currently stored in local storage. If the plugin is loaded without a user, local storage is wiped (for privacy). This means that a user's settings are lost if they log out, switch browsers, clear their browser, switch devices, or if the API is temporarily unreachable.

This PR persists a user's plugin settings as an entity, so that they are preserved and reloaded on all browsers/devices/across log-outs.

It also refactors the order in which various system entities are created in the migration files, so that it's possible in migration scripts to use e.g. machine web actors to update existing entities to new type versions. Previously this was not possible because various system entities were created after all migrations ran, meaning that functions that depended on those entities couldn't be used in the migration.

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:

- [ ] modifies the plugin, which will need publishing once it's approved

### 📜 Does this require a change to the docs?




The changes in this PR:

- [x] adds developer-facing docs to the migrations folder

### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## 🛡 What tests cover this?



- None (I will be adding tests for the plugin soon)

## ❓ How to test this?



1. Load the plugin and set some configuration for manual / automatic inference, write a quick note (without saving)
2. Check that these settings are preserved, e.g.
   - kill the API and open the popup so that local storage is wiped, then restart the API and check the settings reload
   - open the plugin in a different browser or on a different device
